### ISSUE TITLE
[BUG] searchd - too many open files

### ISSUE THREAD

**Issue by dilyanpalauzov** [jsuto/piler#123]
I have Piler 1.4.6 with manticore 6.3.2.  From an old Piler installation I `pilerexport `  100 000 files in a directory.  Then I do on the new installation `pilerimport -d <that-directory>`.  The CPU, used by searchd, increased for long time.

To see what is going on there, I called `strace -p <searchd-PID>` and then it prints
```
accept(7, …) = -1 EMFILE (Too many open files)
accept(8, …) = -1 EMFILE (Too many open files)
```
the above repeat.  /proc/PID/fd/7 and 8 point to socket and the process has opened indeed descriptors from 0 to 1023, all pointing to some socket opened in the last minute.

I do not know, if this “Too many open files” error, returned by accept, is transient and manticore will recover, or if it is forever.

Syslog prints "ERROR: can't connect to 127.0.0.1:9306”.

The problem report is, that `pilerimport -d` shoud stay attached to the terminal, and provide either error message, if searchd gets too many emails, or print success, once everything is imported.

**Comment by jsuto** [comment]
pilerimport prints mostly its output to syslog where you can see any potential issues.

Anyway, can you reorganize the 100k files to 100 dirs having only 1k files? Also it's good to know which process opens that many files. Check lsof output as well.

**Comment by dilyanpalauzov** [comment]
piler/manticore/manticore.log contains:
```
[Wed Jul 31 11:52:37.012 2024] [821] WARNING: last message repeated 100 times
[Wed Jul 31 11:52:37.012 2024] [821] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
```
manticore.log is now 18GB.

syslog contains
```
Jul 31 12:07:08 ip-10-65-6-149 piler[1635]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:08 ip-10-65-6-149 piler[1633]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1624]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1630]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1628]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1637]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1620]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[557]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[554]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1617]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1636]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1632]: ERROR: cant connect to 127.0.0.1:9306
Jul 31 12:07:09 ip-10-65-6-149 piler[1638]: ERROR: cant connect to 127.0.0.1:9306
```
> Also it's good to know which process opens that many files. Check lsof output as well.

It is the searchd process.  What do you want to see from lsof?

**Comment by dilyanpalauzov** [comment]
> Anyway, can you reorganize the 100k files to 100 dirs having only 1k files? 

I could.  But in fact I want to import millions of files, 100k was just a try.  As `pilerimport` is asynchronous, how can I know if the import is ready and I can start with the next 1k files?  Do you have ready code how to split a directory with many files into directories of 1k files?

Also, when emails arrive over SMTP to Piler, it puts in metadata .from and .fromdomain columns the information from the SMTP envelope sender.  Is my understanding correct, that`pilerexport-v1.1.1` creates an eml file.  When it is consumed by `pilerimport-v1.4.6` the original SMTP envelope information is lost and instead the recipient and sender are determined by the MIME-headers From: and To:?

**Comment by jsuto** [comment]
I don't have such code, however, pilerimport runs recursively, you can run it in the top dir, and it will process all subdirs.

You are right, pilerexport doesn't provide smtp envelope information.

**Comment by dilyanpalauzov** [comment]
> pilerimport runs recursively, you can run it in the top dir, and it will process all subdirs.

Does the above mean, that if I create main directory and run `pilerimport -d <that-directory>`, having in each subdirectory 1000 emails, searchd will run again out of file descriptors, as the complete set of emails passed to pilerimport, directly or indirectly (recursively), is above 10 000?

**Comment by jsuto** [comment]
I'm not sure, I suggest to try it. Btw. if you specify -r to pilerimport, then it removes the processed files.

**Comment by dilyanpalauzov** [comment]
For a sample with 27646 files use
```sh
for i in $(find .); do
    if [ ${#i} -gt 4 ]; then
       mkdir ${i:0:-4}
       mv $i ${i:0:-4}/
    fi
done
```
to create one directory per file.  Then I call `pilerimport -r -d .`.

As the suggestion was to split the files in directories of up to 1000 messages, the same problem would be solved if I create more directories with one message in each directory.  Then import the files recursively.

Then piler/manticore/manticore.log starts filling with
```
[Thu Aug  1 13:58:06.090 2024] [873] WARNING: last message repeated 100 times
[Thu Aug  1 13:58:06.090 2024] [873] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
```

**Comment by dilyanpalauzov** [comment]
So, I did the following: I deleted all files from piler/manticore/  and restarted the system.  In the mysql database I could not find any imported emails, but I am not 100% I looked at the correct place.  I assume that at this moment the system does not try to import any files. Then piler/manticore/manticore.log stared filling with the text below.  Thus the out of file descriptors error happens, even if no files are imported or delivered over SMTP. 
```
[Thu Aug  1 14:40:12.155 2024] [922] watchdog: main process 923 forked ok
[Thu Aug  1 14:40:12.164 2024] [923] Using local time zone '/etc/localtime'
[Thu Aug  1 14:40:12.165 2024] [923] starting daemon version '6.3.2 c296dc7c8@24062606 (columnar 2.3.0 88a01c3@24052206) (secondary 2.3.0 88a01c3@24052206) (knn 2.3.0 88a01c3@24052206)' ...
[Thu Aug  1 14:40:12.165 2024] [923] listening on 127.0.0.1:9312 for sphinx and http(s)
[Thu Aug  1 14:40:12.165 2024] [923] listening on 127.0.0.1:9306 for mysql
[Thu Aug  1 14:40:12.165 2024] [923] listening on 127.0.0.1:9307 for RO mysql
[Thu Aug  1 14:40:12.261 2024] [931] prereading 3 tables
[Thu Aug  1 14:40:12.261 2024] [931] preread 3 tables in 0.000 sec
[Thu Aug  1 14:40:12.285 2024] [923] accepting connections
[Thu Aug  1 14:40:12.624 2024] [930] [BUDDY] started v2.3.10 '/usr/share/manticore/modules/manticore-buddy/bin/manticore-buddy --listen=http://127.0.0.1:9312 --bind=127.0.0.1  --threads=2 --skip=manticoresoftware/buddy-plugin-sharding --skip=manticoresoftware/buddy-plugin-queue' at http://127.0.0.1:44523
[Thu Aug  1 14:40:12.624 2024] [930] [BUDDY] Loaded plugins:
[Thu Aug  1 14:40:12.624 2024] [930] [BUDDY]   core: empty-string, backup, emulate-elastic, create, insert, alias, select, show, cli-table, plugin, test, alter-distributed-table, alter-rename-table, modify-table, knn, replace
[Thu Aug  1 14:40:12.624 2024] [930] [BUDDY]   local: 
[Thu Aug  1 14:40:12.624 2024] [930] [BUDDY]   extra: 
[Thu Aug  1 14:44:20.786 2024] [989] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
[Thu Aug  1 14:44:20.789 2024] [989] WARNING: last message repeated 100 times
[Thu Aug  1 14:44:20.789 2024] [989] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
[Thu Aug  1 14:44:20.792 2024] [989] WARNING: last message repeated 100 times
[Thu Aug  1 14:44:20.792 2024] [989] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
[Thu Aug  1 14:44:20.794 2024] [989] WARNING: last message repeated 100 times
[Thu Aug  1 14:44:20.794 2024] [989] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
[Thu Aug  1 14:44:20.796 2024] [989] WARNING: last message repeated 100 times

```

**Comment by dilyanpalauzov** [comment]
For completeness I am mentioning that there is no local mysql server.  config-site.php contains `$config['DB_HOSTNAME'] = 'the-correct-value';` and the web interface shows under Administration → User Management (index.php?route=user/list ) the correct users from the remote database (5.5.5-10.5.25-MariaDB).  The server name is not mentioned anywhere. In manticore.conf I have
```
define('RT', 1);
…
<?php if(RT == 0) { ?>

source base
{
   type = mysql
   sql_host = MYSQL_HOSTNAME
   sql_db = MYSQL_DATABASE
   sql_user = MYSQL_USERNAME
   sql_pass = MYSQL_PASSWORD

   sql_attr_uint = size
   sql_attr_uint = sent
   sql_attr_uint = attachments
}
```
`sql_host = MYSQL_HOSTNAME` should be irrelevant, as `RT` is 1.

**Comment by dilyanpalauzov** [comment]
> The server name is not mentioned anywhere. 

The server name is not mentioned anywhere else, except once in config-site.php.

**Comment by dilyanpalauzov** [comment]
Apart from having remote mysql database, I also had a non-default storage (/data instead of /var). In /etc/manticore.conf I adjusted `define('LOCALSTATEDIR', '/var');`. And then I found the final piece needed: In piler.conf I set mysqlhost, removed mysqlsocket, and added sphinxdir=/data/piler/manticore. (workdir=/data/piler/tmp and queuedir=/data/piler/store were also there).

So the fix was setting mysqlhost= in piler.conf correctly.

**Comment by dilyanpalauzov** [comment]
After configuring piler.conf:pilerhost the import worked for a while, but at some moment
```
[Thu Aug  1 14:44:20.786 2024] [989] WARNING: accept() failed, raise ulimit -n and restart searchd: Too many open files (on linux see /etc/security/limits.conf, 'ulimit -n', max_open_files config option)
[Thu Aug  1 14:44:20.789 2024] [989] WARNING: last message repeated 100 times
```
started again appearing. At some moment I restarted the system.

Can the reason be, that the TCP connection between the two systems (Piler, MySQL) dropped and afterwards Piler has not tried to reestablish it?  Or Piler/searchd does try to establish the connection, but even then does not recover from the out of file descriptors problem?

**Comment by jsuto** [comment]
Not sure. I think it's worth to ask this issue at the manticore support forum. Can you do that?

**Comment by dilyanpalauzov** [comment]
It will be difficult for me to ask this question there, as I do not understand how Piler works with MySQL and Manticore.  Actually, until now I thought Piler connects to MySQL, Piler connects to Manticore, Manticore emulates MySQL interface, but Manticore does not connect to MySQL.

**Comment by dilyanpalauzov** [comment]
> Or Piler/searchd does try to establish the connection, but even then does not recover from the out of file descriptors problem?

I get this error often:
> Aug  5 19:40:37 ip-10-65-6-149 piler[2049]: WARNING: MYSQL_OPT_RECONNECT is deprecated and will be removed in a future version.

and the code contains
> src/mysql.c:19:   mysql_options(&(sdata->mysql), MYSQL_OPT_RECONNECT, (const char*)&rc);
> src/mysql.c:43:   mysql_options(&(sdata->sphx), MYSQL_OPT_RECONNECT, (const char*)&rc);

So I guess in the past mysql did reconnect, when the connection failed, but today MYSQL_OPT_RECONNECT … is deprected and will be removed in a future version.

**Comment by dilyanpalauzov** [comment]
It turned out, that I had in pilersmtp.service, piler.service, or both the line `RuntimeDirectory=piler` together with `User=root`. At the same time pilersearch.service has User=piler, RuntimeDirectory=piler.  This resulted race condititon, where /var/run/piler/ was sometimes created as user piler, sometimes as user root, or something similar.  Eventually the service, which starts searchd, could not create a pid file, systemd thought it never started, and systemd restarted it.  But that service has not stopped its children, but was nevertheless restarted.  Then restarted again, again restarted.

I think this led to a situation, where subsequent processes could not bind to port 9306, so they printed an error, and probably also to too many processes, connecting to mysql.

As mentioned elsewhere, for me it is unclear, who should create /var/run/piler/.  Moreover, it is unclear, if piler.service must start before or after pilersearch.service.  Likewise it is unclear if pilersearch.service should start before or after piler.service. In fact I think both services do not have any order and this is the problem: in order to let both services write in /var/run/piler/ their PID, first pilersearch.service must be started, it creates /var/run/piler as user piler, where it writes its PID.  Then piler.service is started as user root, and it creates its pid file as user root.

Eventually I added to piler-smtp.service and pilersearch.service the line
```
ExecStartPre=/bin/bash -c "mkdir -p /var/run/piler && chown piler:piler /var/run/piler"
```
and this solved all problems with too high system overload.  Probably in piler-smtp.service this is not needed, but after I added I verified that the system is running stable.

I suggest, that Piler refuses to start children processes, if it cannot write a pid file, and that it refuses to start, if it cannot connect to mysql.  The latter case just leads to steadily increasing system load and it was difficult to find the root cause.
